xknx/knxip/cemi_frame.py

```diff
@@ -188,35 +158,31 @@ def from_knx_data_link_layer(self, cemi) -> int:
         # TPCI (transport layer control information)   -> First 14 bit
         # APCI (application layer control information) -> Last  10 bit
 
-        tpci_apci = cemi[9 + addil] * 256 + cemi[10 + addil]
-
-        try:
-            self.cmd = APCICommand(tpci_apci & 0xFFC0)
-        except ValueError:
-            raise UnsupportedCEMIMessage(
-                "APCI not supported: {:#012b} in CEMI: {}".format(
-                    tpci_apci & 0xFFC0, cemi.hex()
-                )
-            )
-
-        apdu = cemi[10 + addil :]
-        if len(apdu) != self.mpdu_len:
+        apdu = cemi[9 + addil :]
+        if len(apdu) != (self.mpdu_len + 1):
             raise CouldNotParseKNXIP(
                 "APDU LEN should be {} but is {} in CEMI: {}".format(
                     self.mpdu_len, len(apdu), cemi.hex()
                 )
             )
 
-        if len(apdu) == 1:
-            apci = tpci_apci & DPTBinary.APCI_BITMASK
-            self.payload = DPTBinary(apci)
-        else:
-            self.payload = DPTArray(cemi[11 + addil :])
+        tpci_apci = (apdu[0] << 8) + apdu[1]
+
+        try:
+            self.payload = APCI.resolve_class(tpci_apci & 0x03FF)()
```

Comment:
huch... what is happening here? 😲😂

Reply:
`APCI.resolve_class` takes the relevant bits from TPCI+APCI, and returns a APCI class type. That is then instantiated directly, so that `self.payload` is one of the APCI classes.

Further down, `self.payload.from_knx(..)` is used to populate it.

Reply:
Could we instantiate it right from resolve_class if we pass the whole adpu? Like a Factory?

We wouldn't need the empty intermediary GroupValue* Class (before from_knx() ) and the somewhat irritating (at least for me) `)()` instantination.

Reply:
Last commit should address this issue. I renamed the function to `resolve_apci` now it doesn't resolve to a type anymore.